operator/internal/controller/vectorized/cluster_controller.go

```diff
@@ -233,11 +233,11 @@ func (r *ClusterReconciler) Reconcile(
 	}
 
 	result, errs := ar.Ensure()
-	if !result.IsZero() && errs == nil {
-		return result, nil
-	}
 	if errs != nil {
-		return result, errs
+		return ctrl.Result{}, errs
+	}
+	if !result.IsZero() {
```

Comment:
This implies that `nil, nil` (well conceptually  `zero, nil`) is a valid result from` .Ensure`. That's _generally_ undesirable and considered unidiomatic. Mind adding a note above the call to Ensure or perhaps a comment on the Ensure function / interface to make this implication more explicit?

Reply:
hm, zero, nil is definitely a valid result of Ensure. in that case, everything went well (or am i missing something?)

Reply:
Ah, sorry I more so mean in the general case of a function that returns `(*T, error)`. Having `(nil, nil)`, `(nil, fmt.Errorf("some error"))`, and `(&T{}, nil)` is considered bad practice as most functions with an error return have two states to handle: success or failure.

This case highlights the exact oddity where there are 3 states that need to be handled instead of just two. As special handling is required, that should be explicitly noted in the code somewhere to ensure this nuance isn't overlooked in the future.